server/src/main/java/org/opensearch/search/DefaultSearchContext.java

```diff
@@ -905,20 +919,74 @@ public boolean shouldUseConcurrentSearch() {
         return concurrentSearchSettingsEnabled && Boolean.TRUE.equals(requestShouldUseConcurrentSearch.get());
     }
 
+    private boolean evaluateAutoMode() {
+
+        // filter out deciders that want to opt-out of decision making
+        Set<ConcurrentSearchDecider> filteredDeciders = concurrentSearchDeciders.stream()
+            .filter(concurrentSearchDecider -> concurrentSearchDecider.canEvaluateForIndex(indexService.getIndexSettings()))
+            .collect(Collectors.toSet());
+        // evaluate based on concurrent search query visitor
+        if (filteredDeciders.size() > 0) {
+            ConcurrentSearchVisitor concurrentSearchVisitor = new ConcurrentSearchVisitor(
+                filteredDeciders,
+                indexService.getIndexSettings()
+            );
+            if (request().source() != null && request().source().query() != null) {
+                QueryBuilder queryBuilder = request().source().query();
+                queryBuilder.visit(concurrentSearchVisitor);
+            }
+        }
+        List<ConcurrentSearchDecision> decisions = filteredDeciders.stream()
```

Comment:
@sohami 
I think we should also have a null check for Decision because some decider might return null if they dont maintain the state properly within the plugin. Null check is good to sanity check the decision given by plugin